core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,210 @@
+   */
+  // Discussion point, make fluent?
+  public abstract void attach();
+
+
+  /**
+   * Close this context and optionally provide a cause for the closure. If this context
+   * cascades closure then notify all child contexts of closure.
+   *
+   * @throws java.lang.IllegalStateException if {@link #close()} has already been called
+   */
+  // Discussion point: Allow for calls when already closed
+  public abstract void close(@Nullable Throwable cause);
+
+  /**
+   * Close the context normally.
+   */
+  @Override
+  public void close() {
+    close(null);
+  }
+
+  /**
+   * Has this context been closed.
+   */
+  public abstract boolean isClosed();
+
+  /**
+   * Return the cause of the closure or {@code null} if context closed without a cause.
+   */
+  @Nullable
+  public abstract Throwable cause();
+
+
+  public Runnable wrap(final Runnable r) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        attach();
+        try {
+          r.run();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  public <C> Callable<C> wrap(final Callable<C> c) {
+    return new Callable<C>() {
+      @Override
+      public C call() throws Exception {
+        attach();
+        try {
+          return c.call();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  public ExecutorService wrap(final ExecutorService e) {
+    // TODO
+    return null;
+  }
+
+  public ScheduledExecutorService wrap(final ScheduledExecutorService e) {
+    // TODO
+    return null;
+  }
+
+  // Discussion point, for Java 6+ would be nice to be able to wrap ForkJoinPool/Task
+  // and some of the newer stuff in Java8 too.
+
+  /**
+   * Create a builder for a new context object with this as its parent. By default the
+   * returned builder uses the cascading closure setting from the parent context.
+   * @param key
+   * @param value
+   * @return
+   */
+  public abstract Builder buildChild();
+
+
+  public static interface Builder {
+
+    /**
+     * Control where a context propagates closure to it's children.
+     */
+    public abstract Builder cascadeClose(boolean enable);
+
+    /**
+     * Add a listener that will be notified when the context is closed.
+     */
+    public abstract Builder addListener(Listener listener);
+
+    /**
+     * Add an value to the context that overrides any value in the parent for the same key.
+     */
+    public abstract <T> Builder addEntry(Key<T> key, T value);
+
+    /**
+     * Build a new context.
+     */
+    public Context build();
+  }
+
+  /**
+   *
+   * @param <T>
+   */
+  public static class Key<T> {
+
+    private final String name;
+    private final T defaultValue;
+
+    public Key(String name) {
+      this(name, null);
+    }
+
+    public Key(String name, T defaultValue) {
+      this.name = Preconditions.checkNotNull(name);
+      this.defaultValue = defaultValue;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T get() {
```

Comment:
We would need to make it _very_ clear that `Context.get()` does not use the default value whereas `Key.get()` does.